### PR TITLE
fix(resolve-template): avoid quoted commas while splitting arguments

### PR DESCRIPTION
**Pull request by glepur**
https://github.com/nxtedition/nxt/issues/3257

### CODE REVIEW

**Comment by ronag** [comment]
What about:

```
 expect(await resolveTemplate(`{{test | join(',') | replace("foo", "bar") | split(",")}}`, { test: ['foo', 'bar'] })).toEqual(['bar', 'bar'])
```

**Comment by glepur** [comment]
Oh ok, I haven't noticed we support those. Update is on the way.

**Comment by glepur** [comment]
Better?

File: src/util/template/expression.js [hunk 1]
```diff
@@ -247,13 +247,13 @@ module.exports = ({ ds } = {}) => {
     const [ , filterName, argsStr ] = filterStr.match(/([^(]+)\((.*)\)/) || []
 
     const tokens = argsStr
-      ? argsStr.match(/(".*?"|[^",\s]+)(?=\s*,|\s*$)/g)
+      ? argsStr.match(/(["`'].*?["`']|[^["`'],\s]+)(?=\s*,|\s*$)/g)
```
Comment (ronag) [inline src/util/template/expression.js]: [^["`'],\s] => [^"`',\s]

Comment (glepur) [reply src/util/template/expression.js]: image

File: src/util/template/expression.js [hunk 2]
```diff
@@ -247,13 +247,13 @@ module.exports = ({ ds } = {}) => {
     const [ , filterName, argsStr ] = filterStr.match(/([^(]+)\((.*)\)/) || []
 
     const tokens = argsStr
-      ? argsStr.split(/\s*,\s*/)
+      ? argsStr.match(/(["`'].*?["`']|[^"`',\s]+)(?=\s*,|\s*$)/g)
```
Comment (ronag) [inline src/util/template/expression.js]: Probably fine. But can you explain this? :D

File: src/util/template/expression.js [hunk 3]
```diff
@@ -247,13 +247,13 @@ module.exports = ({ ds } = {}) => {
...
       : []
 
     const args = tokens
       .map(x => {
         try {
-          return x ? JSON5.parse(x) : x
+          return x ? JSON5.parse(x.replace(/^`(.*)`$/, '\'$1\'')) : x
```
Comment (ronag) [inline src/util/template/expression.js]: This looks suspicious. Why is it needed?

File: src/util/template/expression.js (hunk 2 above)
Comment (glepur) [reply src/util/template/expression.js]: This matches anything inside `` '"` ``, including those characters themselves, separated by comma surrounded by one or more whitespaces.
To be more precise:
&nbsp;&nbsp;&nbsp;&nbsp;`(`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`` ["`'].*?["`'] ``&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`` '"` `` + anything but `` '"` `` + `` '"` ``
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`|`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;or
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;``[^"`',\s]+``&nbsp;&nbsp;&nbsp;&nbsp;1 or more characters excl. `` '"` ``, comma or whitespace
&nbsp;&nbsp;&nbsp;&nbsp;`)`
&nbsp;&nbsp;&nbsp;&nbsp;`(?=`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;followed by
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`\s*,`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0 or more empty spaces and a comma
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;or
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`\s*$`&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0 or more empty spaces and nothing else (end of string)
&nbsp;&nbsp;&nbsp;&nbsp;`)`

File: src/util/template/expression.js (hunk 3 above)
Comment (glepur) [reply src/util/template/expression.js]: `JSON5` does not support backticks, so I simply replace them with single quotes.

File: src/util/template/expression.js (hunk 2 above)
Comment (deadbeef84) [reply src/util/template/expression.js]: If we want to have quotes in our strings, we should probably use back-references for the quotes to make sure they match:
``(["`']).*?\1``.

This still doesn't support backslash escaping such as `"foo\"bar"`, for partial support you can do:
``(["`']).*?(?<!\\)\1``

Only problem with this is ending a string with an escaped backslash, such as `"foo\\"` - bonus points if you fix this :100: but don't spend time on it :rofl: Regex are not really suited for parsing :-1:

File: src/util/template/expression.js (hunk 3 above)
Comment (deadbeef84) [reply src/util/template/expression.js]: Do we really want to support them if we don't support expressions inside?

File: src/util/template/expression.js (hunk 2 above)
Comment (glepur) [reply src/util/template/expression.js]: Haha, yeah. Do we really want to see how deep the rabbit hole goes? 😆

File: src/util/template/expression.js (hunk 3 above)
Comment (glepur) [reply src/util/template/expression.js]: Robert suggested it on his first comment in this issue.
I don't mind removing it, we just have to make sure we will never use them.

Comment (ronag) [reply src/util/template/expression.js]: no backsticks then :) fine with me but we need ' and "

File: src/util/template/expression.js (hunk 2 above)
Comment (ronag) [reply src/util/template/expression.js]: is there no npm package for this? e.g. balanced-match

Comment (glepur) [reply src/util/template/expression.js]: Not that I know of. balanced-match would not work, it just returns everything between given characters...

Comment (glepur) [reply src/util/template/expression.js]: I can put more time into searching for npm package, or write this without regex in worst case.
But I don't think it's necessary to change anything, current solution works fine, don't you think?

Comment (ronag) [reply src/util/template/expression.js]: @glepur I'll have to trust you two on that. That regex is a bit over my head :).

Comment (ronag) [reply src/util/template/expression.js]: https://www.npmjs.com/package/split-string ?

Comment (ronag) [reply src/util/template/expression.js]: `split('a.b,c', { separator: ',', brackets: true,quotes: ['"''] })`

**Comment by ronag** [comment]
Ok, thinking about this deeper, I think we can only support either " or ', not both, otherwise we get ambiguity in the JSON5 parse. Since we already use " out in the wild I think we'll stick with that

File: src/util/template/expression.js (hunk 2 above)
Comment (glepur) [reply src/util/template/expression.js]: I agree regex readability sucks. Thumbs up for that lib, I will try it and see how tests behave.

**Comment by glepur** [comment]
Are you still thinking this through, or you want me to remove `'` right now?

**Pull request closed without merging**